h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -581,50 +609,40 @@ private void learn() {
   }
 
 
-  // There are per (possibly concurrent) AutoML run.
-  // All created keys for a run use the unique AutoML run timestamp, so we can't have name collisions.
-  AtomicInteger individualModelsTrained = new AtomicInteger();
-  private NonBlockingHashMap<String, Integer> algoInstanceCounters = new NonBlockingHashMap<>();
-  private NonBlockingHashMap<String, Integer> gridInstanceCounters = new NonBlockingHashMap<>();
-
-  private int nextInstanceCounter(String algoName, NonBlockingHashMap<String, Integer> instanceCounters) {
-    synchronized (instanceCounters) {
-      int instanceNum = 1;
-      if (instanceCounters.containsKey(algoName))
-        instanceNum = instanceCounters.get(algoName) + 1;
-      instanceCounters.put(algoName, instanceNum);
-      return instanceNum;
+  private int nextInstanceCounter(String algoName, String type) {
+    String key = algoName+"_"+type;
+    if (!_instanceCounters.containsKey(key)) {
+      synchronized (_instanceCounters) {
+        if (!_instanceCounters.containsKey(key))
+          _instanceCounters.put(key, new AtomicInteger(0));
+      }
     }
+    return _instanceCounters.get(key).incrementAndGet();
   }
 
-  Key<Model> modelKey(String algoName, boolean with_counter) {
-    String counterStr = with_counter ? "_" + nextInstanceCounter(algoName, algoInstanceCounters) : "";
-    return Key.make(algoName + counterStr + "_AutoML_" + timestampFormatForKeys.format(_startTime));
-  }
-
-  Key<Grid> gridKey(String algoName, boolean with_counter) {
-    String counterStr = with_counter ? "_" + nextInstanceCounter(algoName, gridInstanceCounters) : "";
-    return Key.make(algoName + "_grid_" + counterStr + "_AutoML_" + timestampFormatForKeys.format(_startTime));
+  Key makeKey(String algoName, String type, boolean with_counter) {
+    String counterStr = with_counter ? "_" + nextInstanceCounter(algoName, type) : "";
+    String prefix = StringUtils.isNullOrEmpty(type) ? algoName : algoName+"_"+type+"_";
+    return Key.make(prefix + counterStr + "_AutoML_" + timestampFormatForKeys.format(_startTime));
```

Comment:
standardized all this `makeKey` logic so that we can always use the same methods for all steps.